### PR TITLE
Optim-wip: Fix bug with nn.Sequential targets

### PR DESCRIPTION
**Pull request by ProGamerGov**
Some PyTorch modules that can be used as targets have an `__iter__` attribute, like `torch.nn.Sequential`. The current target system only checks for the `__iter__` attribute to determine if a target is already in a list or not, and thus doesn't work with targets like Sequential modules. This PR fixes this bug.

This PR also corrects `InputOptimization`'s model variable type hint to reflect the fact that it's optional.

### CODE REVIEW

File: captum/optim/_core/loss.py [hunk 1]
```diff
@@ -128,8 +128,16 @@ def loss_fn(module: ModuleOutputMapping) -> torch.Tensor:
...
-            self.target if hasattr(self.target, "__iter__") else [self.target]
-        ) + (other.target if hasattr(other.target, "__iter__") else [other.target])
+            self.target
+            if hasattr(self.target, "__iter__")
+            and not isinstance(self.target, nn.Module)
+            else [self.target]
+        ) + (
+            other.target
+            if hasattr(other.target, "__iter__")
+            and not isinstance(other.target, nn.Module)
```
Comment (aobo-y) [inline captum/optim/_core/loss.py:138]: This of course will work! But we have already clearly defined the type of `self.target` as l`Union[nn.Module, List[nn.Module]]`. So we only need to check if it is a `list` and no need to worry about other `iterable`. We can just do
```py
if isinstance(self.target, list)
```

Comment (ProGamerGov) [reply captum/optim/_core/loss.py]: Yeah, I think that the `hasattr(obj, __iter__)` is a holdover from when SK was putting together the first iteration of the comprehension system, when weren't 100% sure how it was going to work.

I've replaced the relevant lines with `isinstance(self.target, list)`!

File: captum/optim/_core/optimization.py [hunk 2]
```diff
@@ -56,15 +56,17 @@ def __init__(
                         optimization.
             lr (float): The learning rate to use with the Adam optimizer.
         """
-        self.model = model
+        self.model = model or nn.Identity()
```
Comment (aobo-y) [inline captum/optim/_core/optimization.py:59]: add `optional` to the type in the corresponding doc (comments above) as well? btw, what is the use case of not giving a `model`

File: captum/optim/_core/optimization.py [hunk 3]
```diff
@@ -56,15 +56,17 @@ def __init__(
...
-        if hasattr(loss_function.target, "__iter__"):
+        assert hasattr(loss_function, "target")
+        if isinstance(loss_function.target, list):
             self.hooks = ModuleOutputsHook(loss_function.target)
         else:
             self.hooks = ModuleOutputsHook([loss_function.target])
         self.input_param = input_param or NaturalImage((224, 224))
         if isinstance(self.model, Iterable):
-            self.input_param.to(next(self.model.parameters()).device)
+            if list(self.model.parameters()) != []:
```
Comment (aobo-y) [inline captum/optim/_core/optimization.py:68]: covert the generator to a list means we have to iterate the entire generator and store them, which is unnecessarily expensive. If you just want to check if the model has parameters, try call `next` once with fallback
```py
param = next(self.model.parameters(), None)
if param:
  self.input_param.to(param.device)
```

Comment (ProGamerGov) [reply captum/optim/_core/optimization.py]: @aobo-y Thanks for the tip! I've changed the code to your suggested less intensive way of checking for parameters to use as a source for what device to use.

File: captum/optim/_core/optimization.py (hunk 2 above)
Comment (ProGamerGov) [reply captum/optim/_core/optimization.py:58]: @aobo-y I thought that listing optional in the docs meant that there was a preset value for the variable? I've added it now though.

The reasoning behind making the model optional is that for example in some cases the image parameterization may be a generative model. Thus it doesn't make a lot of sense to require a second model be used in these cases unless the user wants to link them.
